### PR TITLE
Add support for rspec3

### PR DESCRIPTION
**Pull request by Intrepidd**
There was several issues :
- `with_around_each_hooks` is renamed to `with_around_example_hooks` in rspec 3.0
- `@example` does not exist anymore in `__run_next_step__` This issue was present in rspec 2.99 as well , it made all specs pass even if they should not. Instead, I clean the steps and re-raise the exception that is later caught by rspec himself. Not sure about that exactly, but I tested with rspec 2.14 and everything seems fine so backwards compatibility seems to be in the party.

Cheers

### CODE REVIEW

File: lib/rspec/eventmachine/async_steps.rb [hunk 1]
```diff
@@ -1,16 +1,16 @@
 module RSpec::EM
   class AsyncSteps < Module
-
```
Comment (jcoglan) [inline lib/rspec/eventmachine/async_steps.rb:3]: Could you get all the whitespace changes out of this commit? Formatting changes should be separated from functional changes or it becomes very hard to read the history.

File: lib/rspec/eventmachine/async_steps.rb [hunk 2]
```diff
@@ -28,47 +28,47 @@ def __enqueue__(args)
...
-        
+
         method_name, args = step.shift, step
         begin
           method(method_name).call(*args) { __run_next_step__ }
-        rescue Object => error
-          @example.set_exception(error) if @example
+        rescue Object
           __end_steps__
+          raise
```
Comment (jcoglan) [inline lib/rspec/eventmachine/async_steps.rb:41]: What's going on here? You're calling `raise` with no argument; how is the actual exception reported to the runner?

**Comment by jcoglan** [comment]
I've left some notes on the code on the diff itself. As a side note, I would prefer that you not use gendered pronouns to refer to software on my repos. RSpec is 'itself', not 'himself'.

File: lib/rspec/eventmachine/async_steps.rb (hunk 1 above)
Comment (Intrepidd) [inline lib/rspec/eventmachine/async_steps.rb:3]: Sure I'll do that, my editor did it automatically.

## 

Adrien Siami

File: lib/rspec/eventmachine/async_steps.rb (hunk 2 above)
Comment (Intrepidd) [inline lib/rspec/eventmachine/async_steps.rb:41]: Calling raise in a rescue block in Ruby will reraise the same error

## 

Adrien Siami

**Comment by Intrepidd** [comment]
I've answered your comments.

About your side note, pardon my french (literally, we don't have a word for
"it")

## 

Adrien Siami

**Comment by Intrepidd** [comment]
btw I removed the trailing whitespace changes as you asked.

**Comment by Intrepidd** [comment]
`teardown_mocks_for_rspec` and `verify_mocks_for_rspec` should be changed as well, I'll investigate this

**Comment by Intrepidd** [comment]
Ok I improved the PR

I chose to defer run_after_each / run_after_example instead of deferring mock verify / teardown.

This way, code defined in `after(:each)` blocks is guaranteed to be run after EventMachine has finished, including mock verify and teardown.

BTW, it was broken, if you defined the following method in an async step :

``` ruby
def mock_expect &resume
  @foo = double()
  @foo.should_receive :foo
  resume.call
end
```

The tests would pass even if `@foo` did not receive `:foo`

I also updated the travis setup so the gem will be tested against rspec 2.14.x, rspec 2.99.x, rspec 3.x and also rspec edge with allowed failures.

**Comment by jcoglan** [comment]
Running `after` hooks without EventMachine running breaks most of my use cases; `after` hooks are usually used to disconnect clients and stop servers. Is there a compelling reason to drop the mock verification/teardown hooks that I had, or a cleaner way to ensure `after` runs but _inside_ EventMachine?

**Comment by jcoglan** [comment]
I've removed the rescue clauses that rely on `@example` from `{verify,teardown}_mocks_for_rspec` -- see a1d47dd. That change plus your changes seems to be working well with faye-websocket and reporting errors where it's supposed to.

**Comment by Intrepidd** [comment]
Alright, I'll merge your changes into my branch when I reach a computer

## 

Adrien Siami

**Comment by Intrepidd** [comment]
Oh I see you cherry-picked my commits :) I Guess we're good to go then, thanks, I'll let you close it to be sure we're good

**Comment by jcoglan** [comment]
About your `mock_expect` example from above: I've tested mocking in my own code and found it working after these changes. Was it broken because `verify_mocks_for_rspec` relied on `@example`? Does mocking work for you now?

**Comment by Intrepidd** [comment]
​Ok actually it seems that mocking was working ​with rspec 2.14.1, I must
have mixed up rspec versions.

It was not working in rspec 2.99 and rspec 3 because @example was nil though

On Tue, Jun 17, 2014 at 12:15 PM, James Coglan notifications@github.com
wrote:

> About your mock_expect example from above: I've testing mocking in my own
> code and found it working after these changes. Was it broken because
> verify_mocks_for_rspec relied on @example? Does mocking work for you now?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/jcoglan/rspec-eventmachine/pull/3#issuecomment-46289638
> .

## 

Adrien Siami
_Software Engineer_

_https://github.com/Intrepidd https://github.com/Intrepidd_

**Comment by jcoglan** [comment]
I've just released version 0.2.0 with the above patches.

**Pull request closed without merging**